File: frontend/lib/services/sync_manager.dart
import 'package:flutter/cupertino.dart';
import 'package:hive/hive.dart';

import '../models/sync_task.dart';
import 'api/base_api.dart';

class SyncManager {
  final Box<SyncTask> _queueBox = Hive.box<SyncTask>('sync_queue');
  bool _isProcessing = false;

  void enqueue(String method, String path, dynamic body) {
    final task = SyncTask(method: method, path: path, body: body as Map<String, dynamic>?);
    _queueBox.add(task);
  }

  Future<void> processQueue(BaseApi api) async {
    if (_isProcessing || _queueBox.isEmpty) return;
    _isProcessing = true;

    final keys = _queueBox.keys.toList();
    for (var key in keys) {
      final task = _queueBox.get(key);
      if (task == null) continue;

      try {
        await api.request(
          method: task.method,
          path: task.path,
          body: task.body,
          fromJson: (json) => json,
        );
        await _queueBox.delete(key);
      } catch (e) {
        debugPrint("Replay failed for ${task.path}: $e");
        break;
      }
    }
    _isProcessing = false;
  }
}

File: frontend_react/src/services/syncManager.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

export interface SyncTask {
    id: string;
    method: 'POST' | 'PUT' | 'DELETE' | 'PATCH';
    path: string;
    body: Record<string, any> | null;
}

class SyncManager {
    private readonly STORAGE_KEY = 'sync_queue';
    private isProcessing = false;

    /**
     * Enqueues a mutations task to AsyncStorage for offline fallback
     */
    async enqueue(method: SyncTask['method'], path: string, body: any): Promise<void> {
        try {
            const queue = await this.getQueue();

            const newTask: SyncTask = {
                id: Math.random().toString(36).substring(2, 9) + Date.now().toString(),
                method,
                path,
                body: body || null,
            };

            queue.push(newTask);
            await AsyncStorage.setItem(this.STORAGE_KEY, JSON.stringify(queue));
            console.log(`📥 Task queued offline: ${method} ${path}`);
        } catch (error) {
            console.error('❌ Failed to enqueue offline task:', error);
        }
    }

    /**
     * Processes the backlog queue sequentially when connection returns
     */
    async processQueue(apiInstance: any): Promise<void> {
        if (this.isProcessing) return;

        const queue = await this.getQueue();
        if (queue.length === 0) return;

        this.isProcessing = true;
        console.log(`🔄 Processing offline sync queue (${queue.length} tasks)...`);

        const remainingTasks = [...queue];

        for (const task of queue) {
            try {
                await apiInstance.request({
                    method: task.method,
                    path: task.path,
                    body: task.body,
                });

                remainingTasks.shift();
                await AsyncStorage.setItem(this.STORAGE_KEY, JSON.stringify(remainingTasks));
                console.log(`✅ Replay success for: ${task.path}`);
            } catch (error) {
                console.error(`❌ Replay failed for ${task.path}:`, error);
                break;
            }
        }

        this.isProcessing = false;
    }

    /**
     * Helper to retrieve and parse the current queue array safely
     */
    async getQueue(): Promise<SyncTask[]> {
        try {
            const storedQueue = await AsyncStorage.getItem(this.STORAGE_KEY);
            return storedQueue ? (JSON.parse(storedQueue) as SyncTask[]) : [];
        } catch (_) {
            return [];
        }
    }
}

export const syncManager = new SyncManager();